### PR TITLE
rework properties and traits

### PR DESCRIPTION
**Pull request by gilmore606**

### CODE REVIEW

File: src/main/kotlin/world/trait/Trait.kt [hunk 1]
```diff
@@ -25,45 +28,125 @@ sealed class Trait(val name: String) {
...
+    val children: MutableList<ID> = mutableListOf()
 
     val commands: MutableSet<Command> = mutableSetOf()
     val verbs: MutableMap<String, Verb> = mutableMapOf()
-    val props: MutableMap<String, Value> = mutableMapOf()
 
+    // All props we define directly OR inherit.
+    val props: MutableMap<String, Propval> = mutableMapOf()
+
+    // All objects which inherit this trait (directly or transitively).
```
Comment (melville) [inline src/main/kotlin/world/trait/Trait.kt:50]: I'm sure you thought about this, but it seems like keeping track of descendants this way will lead to some very large sets of descendants at the top of the tree.  Would it be better to only track direct descendants in each trait, and then traverse the child traits?

**Review by melville — APPROVED** [review]
Nice work, this is something that is obviously very necessary!  I don't see any obvious problem here, but I'm not super confident that I've considered all the possibilities and edge cases, there is a lot to grok.  I daresay we might benefit from a set of tests that cover all of the possible situations we expect to see in a real DB (chparenting traits in various ways and checking their descendant props).

File: src/main/kotlin/world/trait/Trait.kt (hunk 1 above)
Comment (gilmore606) [reply src/main/kotlin/world/trait/Trait.kt:50]: I did think about it, but that was before I added $root to be honest.  I think you may be right about this.  I only need to look at $trait.objects when we add/remove properties; since we track $trait.children and .parents, we can traverse that to get to all the inheritors when we do those things.  And those ops don't need to be "hot path optimal".

I believe I will make this change.

File: src/main/kotlin/world/Obj.kt [hunk 2]
```diff
@@ -20,58 +22,63 @@ class Obj {
...
-    fun getProp(name: String): Value? {
-        props[name]?.also { return it }
+    // Props
 
-        traits.forEach {
-            Yegg.world.traits[it]?.getProp(this, name)?.also { return it }
-        }
-        return null
+    fun addProp(propName: String, trait: Trait) {
+        if (!props.containsKey(propName)) props[propName] = Propval(trait.id)
```
[the remark this answers is not in the document]
Comment (gilmore606) [reply src/main/kotlin/world/Obj.kt:65]: It is inconsistent, I will add a throw here.

**Comment by gilmore606** [comment]
> I daresay we might benefit from a set of tests that cover all of the possible situations we expect to see in a real DB (chparenting traits in various ways and checking their descendant props).

I think you're right, this is overdue now.  I'm gonna belay finishing the parser etc and focus on a test suite for the next PR.
